src/audio/kpb.c

```diff
@@ -997,18 +997,23 @@ static void kpb_init_draining(struct comp_dev *dev, struct kpb_client *cli)
 
 		} while (buff != first_buff);
 
-		/* Calculate time in clock ticks each draining event shall
-		 * take place. This time will be used to synchronize us with
-		 * an end application interrupts.
+		/* Should we drain in synchronized mode (sync_draining_mode)?
+		 * Note! We have already verified host params during
+		 * kpb_prepare().
 		 */
-
-		drain_interval = (host_period_size / bytes_per_ms) *
-				 ticks_per_ms;
-		/* In draining intervals we will fill only two periods
-		 * and give host time to read it.
-		 * This way we are safe to not overflow host buffer.
-		 */
-		period_bytes_limit = host_period_size * 2;
+		if (kpb->hb_buffer_size > host_buffer_size) {
+			/* Calculate time in clock ticks each draining event
+			 * shall take place. This time will be used to
+			 * synchronize us with application interrupts.
+			 */
+			drain_interval = (host_period_size / bytes_per_ms) *
+					 ticks_per_ms;
+			period_bytes_limit = host_period_size;
```

Comment:
Btw, this part seems ok, but I noted that calculation of 'bytes_per_ms' seems suspicious:

size_t bytes_per_ms = KPB_SAMPLING_WIDTH *
			      (KPB_SAMPLE_CONTAINER_SIZE(sample_width) / 8) *
			      kpb->config.no_channels;

So "bits * bits / 8 * channels", and typically 16*16/8*2=64 How is this bytes per ms? To calculate bytes-per-ms, I'd assume you'd have to factor in the sampling rate, right?

Reply:
> How is this bytes per ms?
So how many bytes do you expect per ms? For 16 kHz sampling and 2 bytes samples and 2 channels?  To me it is 64 bytes of data each millisecond.
However the problem in this formula is bad naming for `KPB_SAMPLING_WIDTH ` macro. It really stands for "number of samples per ms". So I will rewrite this formula to make it more clear.

Reply:
Ack, the answer is ok, but the comments and the formula don't add up. The comment even says:
KPB_SAMPLING_WIDTH 16 /**< number of bits */
... but as you state here, it's "khz" not "bits". This is very confusing even if the numbers end up right. 

Btw, and sorry, this isn't really something you addded in this PR. So can be fixed in another PR as well.

Reply:
@kv2019i, good you spotted it, it's probably the result of some unsuccessful refactor. Anyway I will make it more clear.